dask/array/core.py

```diff
@@ -5651,89 +5651,115 @@ def _vindex_array(x, dict_indexes):
 
-    if points:
-        per_block = groupby(3, points)
-        per_block = {k: v for k, v in per_block.items() if v}
-
-        other_blocks = list(
-            product(
-                *[
-                    list(range(len(c))) if i is None else [None]
-                    for i, c in zip(flat_indexes, x.chunks)
-                ]
-            )
+    if npoints > 0:
+        other_blocks = product(
+            *[
+                list(range(len(c))) if i is None else [None]
+                for i, c in zip(flat_indexes, x.chunks)
+            ]
         )
 
         full_slices = [slice(None, None) if i is None else None for i in flat_indexes]
 
+        # The output is constructed as a new dimension and then reshaped
+        # So the index of the output point is simply an `arange`
+        outinds = np.arange(npoints).reshape(broadcast_shape)
+        # Which output block is the point going to, and what is the index within that block?
+        outblocks, outblock_idx = np.divmod(outinds, max_chunk_point_dimensions)
+
+        # Now we try to be clever. We need to construct a graph where
+        # if input chunk (0,0) contributes to output chunks 0 and 2 then
+        # (0,0) => (0, 0, 0) and (2, 0, 0).
+        # The following is a groupby over output key by using ravel_multi_index to convert
+        # the (output_block, *input_block) tuple to a unique code.
+        ravel_shape = (
+            n_chunks + 1,
+            *(
+                n
+                for i, n in zip(flat_indexes, x.numblocks, strict=True)
+                if i is not None
+            ),
+        )
+        keys = np.ravel_multi_index([outblocks, *block_idxs], ravel_shape)
+        # by sorting the data that needs to be inserted in the graph here,
+        # we can slice in the hot loop below instead of using fancy indexing which will
+        # always copy inside the hot loop.
+        sortidx = np.argsort(keys, axis=None)
+        sorted_keys = keys.flat[sortidx]  # flattens
+        sorted_inblock_idxs = [_.flat[sortidx] for _ in inblock_idxs]
+        sorted_outblock_idx = outblock_idx.flat[sortidx]
+        # Determine the start and end of each unique key. We will loop over this
+        flag = np.concatenate([[True], sorted_keys[1:] != sorted_keys[:-1], [True]])
+        (key_bounds,) = flag.nonzero()
+
         name = "vindex-slice-" + token
         vindex_merge_name = "vindex-merge-" + token
         dsk = {}
         for okey in other_blocks:
             merge_inputs = defaultdict(list)
             merge_indexer = defaultdict(list)
-            for i, key in enumerate(per_block):
+            for i, (start, stop) in enumerate(
+                zip(key_bounds[:-1], key_bounds[1:], strict=True)
+            ):
+                slicer = slice(start, stop)
+                key = sorted_keys[start]
+                outblock, *input_blocks = np.unravel_index(key, ravel_shape)
+                inblock = [_[slicer] for _ in sorted_inblock_idxs]
```

Comment:
we might ask whether we just want to `scatter` `sorted_inblock_idxs` and embed a reference to those along with the slicers in the task instead of the full array

Reply:
Yeah `UserWarning: Sending large graph of size 738.91 MiB.` for

```python
import dask.array as da
import numpy as np
import scipy as sp

chunksize = 100
size = 10_000
n_points = 5000

X = da.random.poisson(15, (size, size), chunks = (chunksize, chunksize))

index_0 = np.random.randint(0, X.shape[0], n_points)
index_0.sort()
index_1 = np.random.randint(0, X.shape[1], n_points)
index_1.sort()

print('vindex timing:')
X.vindex[np.ix_(index_0, index_1)].compute()
```